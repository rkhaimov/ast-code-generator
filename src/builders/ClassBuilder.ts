import reduce from 'lodash/reduce';
import last from 'lodash/last';
import isEmpty from 'lodash/isEmpty';

import { ISwagger, ISwaggerMethod, ISwaggerMethodParam, ISwaggerMethods } from '../definitions/swagger';
import {
  FunctionExpressionBodyTypes,
  IClassDeclaration,
  IFunctionParam,
  IMethodDefinition,
} from '../definitions/class/ast';

interface IClassBuilder {
  build(name: string, methods: ISwagger['paths']): IClassDeclaration;
}

class _ClassBuilder implements IClassBuilder {
  build(name: string, methods: ISwagger['paths']): IClassDeclaration {
    const body = reduce(methods, (acc, operation, api) => {
      const method = this.buildMethod(api, operation);

      return acc.concat(method);
    }, [] as IMethodDefinition[]);

    return {
      type: 'ClassDeclaration',
      id: {
        type: 'Identifier',
        name: `${name}Repository`,
      },
      superClass: {
        type: 'Identifier',
        name: 'RepositoryBase', // TODO: Replace to class name
      },
      body: {
        type: 'ClassBody',
        body,
      },
    };
  }

  private buildMethod(api: string, operations: ISwaggerMethods): IMethodDefinition {
    const method: ISwaggerMethod = 'get' in operations ? operations.get : operations.post;
    const methodName = last<string>(api.split('/'));

    return {
      type: 'MethodDefinition',
      kind: 'method',
      static: false,
      computed: false,
      key: {
        name: methodName!,
        type: 'Identifier',
      },
      value: {
        type: 'FunctionExpression',
        params: this.buildFunctionParams(method.parameters),
        body: {
          type: 'BlockStatement',
          body: this.buildFunctionImplementation(api, operations),
        },
        async: false,
        generator: false,
        expression: false,
        id: null,
      },
    };
  }

  private buildFunctionParams(swaggerParams: ISwaggerMethodParam[]): IFunctionParam[] {
    if (isEmpty(swaggerParams)) {
      return [];
    }

    const param = {
      name: 'payload',
      type: 'Identifier',
    };

    return [param as IFunctionParam];
  }

  private buildFunctionImplementation(api: string, operations: ISwaggerMethods): FunctionExpressionBodyTypes[] {
    return [];
  }
}

export const ClassBuilder: IClassBuilder = new _ClassBuilder();
