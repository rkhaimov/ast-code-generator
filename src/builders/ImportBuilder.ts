import { ImportDeclarationStructure } from 'ts-simple-ast';

import { ProjectManager } from '../managers/ProjectManager';

class _ImportBuilder {
  importRepositoryBase(): ImportDeclarationStructure {
    return {
      moduleSpecifier: `./${ProjectManager.FILE_NAMES.REPOSITORY_BASE}`,
      namedImports: [{ name: ProjectManager.FILE_NAMES.REPOSITORY_BASE }],
    };
  }
}

export const ImportBuilder = new _ImportBuilder();
