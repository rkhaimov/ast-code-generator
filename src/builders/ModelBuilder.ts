/* tslint:disable:max-line-length */
import { reduce, last } from 'lodash';
import { InterfaceDeclarationStructure, PropertySignatureStructure } from 'ts-simple-ast';

import { ISwagger, SwaggerDefinitionPropertyTypes } from '../definitions/swagger';

class _ModelBuilder {
  buildModel(model: string, definition: ISwagger['definitions'][string]): InterfaceDeclarationStructure {
    const properties: InterfaceDeclarationStructure['properties'] = reduce(definition.properties, (acc, value: SwaggerDefinitionPropertyTypes, name) => {
      const property: PropertySignatureStructure = {
        name,
        type: this.toTsType(value),
      };

      return (acc as InterfaceDeclarationStructure['properties'])!.concat(property);
    }, [] as InterfaceDeclarationStructure['properties']);

    return {
      name: `I${model}`,
      properties,
      isExported: true,
    };
  }

  private toTsType(property: SwaggerDefinitionPropertyTypes): string {
    switch (property.type) {
      case 'integer': {
        return 'number';
      }
      case 'string': {
        if (property.enum) {
          return property.enum.map((value) => `'${value}'`).join(' | ');
        }

        return 'string';
      }
      case 'array': {
        if (!('$ref' in property.items)) {
          throw new Error('Unrecognizable array swagger structure');
        }

        const name = this.getRefDefinitionName(property.items.$ref);

        return `${name}[]`;
      }
      default: {
        return property.type;
      }
    }
  }

  private getRefDefinitionName(ref: string): string {
    return `I${last(ref.split('/'))}`;
  }
}

export const ModelBuilder = new _ModelBuilder();
