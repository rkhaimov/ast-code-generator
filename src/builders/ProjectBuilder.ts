export class ProjectBuilder {
  static FILE_NAMES = {
    DEFINITIONS: 'definitions',
    REPOSITORIES: 'repositories',
    REPOSITORY_BASE: 'RepositoryBase',
  };
}
