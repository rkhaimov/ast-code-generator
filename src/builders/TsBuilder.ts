import { ts } from 'ts-simple-ast';

class _TsBuilder {
  project = ts.createSourceFile('ts-build-source.ts', '', ts.ScriptTarget.ES2015, true, ts.ScriptKind.TS);
  printer = ts.createPrinter();

  print(node: ts.Node): string {
    return this.printer.printNode(ts.EmitHint.Unspecified, node, this.project);
  }
}

export const TsBuilder = new _TsBuilder();
