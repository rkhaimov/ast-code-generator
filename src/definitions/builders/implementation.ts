import { ISwaggerMethodParam } from '../swagger';

export type ArgumentsGroup = {
  [P in ISwaggerMethodParam['in']]: ISwaggerMethodParam;
};
