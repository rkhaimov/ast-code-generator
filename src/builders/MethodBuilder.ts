import { isEmpty, last, keys, first, get, lowerFirst, isNil } from 'lodash';
import { MethodDeclarationStructure, ParameterDeclarationStructure, ts } from 'ts-simple-ast';
import escodegen from 'escodegen';

import { GetImplementationBuilder } from './implementation/methods/GetImplementationBuilder';
import { PostImplementationBuilder } from './implementation/methods/PostImplementationBuilder';
import { PutImplementationBuilder } from './implementation/methods/PutImplementationBuilder';

import { BlockStatementBody } from '../definitions/ast/common';
import {
  ISwaggerMethod,
  SwaggerMethodParam,
  ISwaggerOperations,
  SwaggerDefinitionPropertyTypes,
} from '../definitions/swagger';
import { ModelBuilder } from './ModelBuilder';
import { TsBuilder } from './TsBuilder';

interface IMethodBuilder {
  buildMethod(api: string, operations: ISwaggerOperations): MethodDeclarationStructure;
}

class _MethodBuilder implements IMethodBuilder {
  buildMethod(api: string, operations: ISwaggerOperations): MethodDeclarationStructure {
    const methodName = last<string>(api.split('/'));
    const methods: string[] = keys(operations);
    const method: ISwaggerMethod = get(operations, first(methods)!);

    return {
      name: methodName!,
      parameters: this.buildFunctionParams(method.parameters),
      bodyText: this.buildBodyText(api, operations),
      returnType: this.getReturnType(method.responses),
    };
  }

  private getReturnType(response: ISwaggerMethod['responses']): string {
    if (isNil(response['200'].schema)) {
      return TsBuilder.print(ts.createTypeReferenceNode('Promise', [ts.createTypeReferenceNode('void', [])]));
    }

    if ('$ref' in response['200'].schema) {
      const name = ModelBuilder.getRefDefinitionName(response['200'].schema.$ref);

      return TsBuilder.print(ts.createTypeReferenceNode('Promise', [ts.createTypeReferenceNode(name, [])]));
    }

    const type = ModelBuilder.toTsType(response['200'].schema);

    return TsBuilder.print(ts.createTypeReferenceNode('Promise', [ts.createTypeReferenceNode(type, [])]));
  }

  private buildBodyText(api: string, operations: ISwaggerOperations) {
    return escodegen.generate(this.buildFunctionBody(api, operations));
  }

  private buildFunctionBody(api: string, operations: ISwaggerOperations): BlockStatementBody {
    if ('get' in operations) {
      return GetImplementationBuilder.buildImplementation(api, operations.get);
    }

    if ('post' in operations) {
      return PostImplementationBuilder.buildImplementation(api, operations.post);
    }

    if ('put' in operations) {
      return PutImplementationBuilder.buildImplementation(api, operations.put);
    }

    if ('delete' in operations) {
      return PutImplementationBuilder.buildImplementation(api, operations.delete);
    }

    throw new Error(`Unknown operation(s) was provided ${keys(operations)}`);
  }

  private buildFunctionParams(swaggerParams: SwaggerMethodParam[]): ParameterDeclarationStructure[] {
    if (isEmpty(swaggerParams)) {
      return [];
    }

    const literalTypeNodes = swaggerParams.map<ts.TypeElement>((parameter) => {
      const typeName = ModelBuilder.toTsType(this.convertToSwaggerDefinition(parameter));
      const questionToken = parameter.required ? undefined : ts.createToken(ts.SyntaxKind.QuestionToken);

      return ts.createPropertySignature(
        [],
        lowerFirst(parameter.name),
        questionToken,
        ts.createTypeReferenceNode(typeName, []),
        undefined,
      );
    });

    const param: ParameterDeclarationStructure = {
      name: 'payload',
      type: TsBuilder.print(ts.createTypeLiteralNode(literalTypeNodes)),
    };

    return [param];
  }

  private convertToSwaggerDefinition(param: SwaggerMethodParam): SwaggerDefinitionPropertyTypes {
    switch (param.in) {
      case 'body': {
        return {
          type: 'object',
          properties: param.schema,
        };
      }
      case 'path': {
        return {
          ...param as SwaggerDefinitionPropertyTypes,
          type: 'string',
          format: param.format === 'int32' ? undefined : param.format,
        };
      }
      case 'query': {
        return param as SwaggerDefinitionPropertyTypes;
      }
      default: {
        return this.throwIfNever(param, '');
      }
    }
  }

  private throwIfNever(arg: never, message: string): never {
    throw new Error(message);
  }
}

export const MethodBuilder: IMethodBuilder = new _MethodBuilder();
