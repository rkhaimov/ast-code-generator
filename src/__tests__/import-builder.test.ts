import { Project, SourceFile } from 'ts-simple-ast';

import { ImportBuilder } from '../builders/ImportBuilder';

describe('Import builder works well when', () => {
  const project = new Project();
  const file: SourceFile = project.createSourceFile('imports');

  it('matches giving code style', () => {
    file.addImportDeclaration(ImportBuilder.importRepositoryBase());

    expect(file.getFullText()).toMatchSnapshot();
  });
});
