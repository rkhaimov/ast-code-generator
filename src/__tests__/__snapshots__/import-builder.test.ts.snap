// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Import builder works well when matches giving code style 1`] = `
"import { RepositoryBase } from \\"./RepositoryBase\\";
"
`;
